File: java/src/main/java/com/peshchuk/topjslibs/PageJsSources.java
package com.peshchuk.topjslibs;

import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.function.Supplier;
import java.util.regex.Pattern;

public class PageJsSources implements Callable<Set<String>> {

  private static final Pattern SCRIPT_SRC = Pattern
      .compile("<script[^>]+src=[\"']([^\"']+/)?([^\"']+)[\"']");

  private final Supplier<String> htmlSupplier;

  public PageJsSources(final Supplier<String> htmlSupplier) {
    this.htmlSupplier = htmlSupplier;
  }

  @Override
  public Set<String> call() {
    final var html = htmlSupplier.get();
    final var matcher = SCRIPT_SRC.matcher(html);
    final var result = new HashSet<String>();

    while (matcher.find()) {
      if (matcher.groupCount() < 2) {
        continue;
      }

      final var src = matcher.group(2);
      result.add(src);
    }

    return Collections.unmodifiableSet(result);
  }
}

File: rust/src/top_js_libs/page_js_sources.rs
use std::collections::HashSet;

use regex::Regex;

lazy_static! {
    static ref SCRIPT_SRC: Regex = Regex::new(r#"<script[^>]+?src=["'](?:[^"']+/)?([^"']+?)["']"#).unwrap();
}

pub struct PageJsSources<'a> {
    fetch_html: Box<dyn Fn() -> String + 'a>,
}

impl<'a> PageJsSources<'a> {
    pub fn new<'b>(fetch_html: Box<dyn Fn() -> String + 'b>) -> PageJsSources {
        PageJsSources {
            fetch_html
        }
    }

    pub fn get(&self) -> HashSet<String> {
        let html = (self.fetch_html)();
        let mut result = HashSet::new();

        for capture in SCRIPT_SRC.captures_iter(&html) {
            if capture.len() < 2 {
                continue;
            }

            result.insert(capture[1].to_string());
        }

        result
    }
}